src/main/java/com/google/jenkins/plugins/credentials/oauth/JsonServiceAccountConfig.java

```diff
@@ -53,36 +63,109 @@
   private static final long serialVersionUID = 6818111194672325387L;
   private static final Logger LOGGER =
       Logger.getLogger(JsonServiceAccountConfig.class.getSimpleName());
-  private String jsonKeyFile;
+  @CheckForNull
+  private String filename;
+  @CheckForNull
+  private SecretBytes secretJsonKey;
+  @Deprecated   // for migration purpose
+  @CheckForNull
+  private transient String jsonKeyFile;
   private transient JsonKey jsonKey;
 
+  /** @since 0.8 */
   @DataBoundConstructor
-  public JsonServiceAccountConfig(FileItem jsonKeyFile,
-      String prevJsonKeyFile) {
-    if (jsonKeyFile != null && jsonKeyFile.getSize() > 0) {
+  public JsonServiceAccountConfig() {}
+
+  /**
+   * For being able to load credentials created with versions < 0.8
+   * and backwards compatibility with external callers.
+   *
+   * @param jsonKeyFile The uploaded json key file
+   * @param prevJsonKeyFile The path of the previous json key file
+   * @since 0.3
+   */
+  @Deprecated
+  public JsonServiceAccountConfig(
+      FileItem jsonKeyFile, String prevJsonKeyFile) {
+    this.setJsonKeyFileUpload(jsonKeyFile);
+    if (filename == null && prevJsonKeyFile != null) {
+      this.filename = extractFilename(prevJsonKeyFile);
+      this.secretJsonKey = getSecretBytesFromFile(prevJsonKeyFile);
+    }
+  }
+
+  /**@param jsonKeyFileUpload uploaded json key file */
+  @DataBoundSetter // Called on form submit, only used when key file is uploaded
+  public void setJsonKeyFileUpload(FileItem jsonKeyFileUpload) {
+    if (jsonKeyFileUpload != null && jsonKeyFileUpload.getSize() > 0) {
       try {
         JsonKey jsonKey = JsonKey.load(new JacksonFactory(),
-                jsonKeyFile.getInputStream());
+            jsonKeyFileUpload.getInputStream());
         if (jsonKey.getClientEmail() != null &&
-                jsonKey.getPrivateKey() != null) {
-          try {
-            this.jsonKeyFile = writeJsonKeyToFile(jsonKey);
-          } catch (IOException e) {
-            LOGGER.log(Level.SEVERE, "Failed to write json key to file", e);
-          }
+            jsonKey.getPrivateKey() != null) {
+          this.filename = extractFilename(jsonKeyFileUpload.getName());
```

Comment:
It seems like these two variables are linked by usage. Does it make sense combining them into a struct?

Reply:
After seeing how they are set, this probably doesn't make sense. Deleting.